### PR TITLE
[INFRA/CORE] Update only client's own info in GameState

### PR DESCRIPTION
**Pull request by lhl2617**
# Summary

As title. We don't want all clients to know each other's info.

## Additional Information

Removed action code for good. Messes up the structure too much.

## Test Plan

- `go test ./...`
```
?       github.com/SOMAS2020/SOMAS2020  [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team1   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team2   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team3   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team4   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team5   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team6   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/common/baseclient       [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/common/config   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/common/gamestate        [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/common/rules    (cached)
?       github.com/SOMAS2020/SOMAS2020/internal/common/shared   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/logger  [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/server  0.276s
ok      github.com/SOMAS2020/SOMAS2020/pkg/fileutils    (cached)
?       github.com/SOMAS2020/SOMAS2020/pkg/miscutils    [no test files]
?       github.com/SOMAS2020/SOMAS2020/pkg/testutils    [no test files]
```

- `go run .`
```
2020/12/20 16:09:28 [SERVER]: start runTurn
2020/12/20 16:09:28 [SERVER]: TURN: 1, Season: 1
2020/12/20 16:09:28 [SERVER]: start startOfTurnUpdate
2020/12/20 16:09:28 [Team2]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x1, ClientInfo:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team4]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x1, ClientInfo:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team6]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x1, ClientInfo:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team1]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x1, ClientInfo:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team3]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x1, ClientInfo:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team5]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x1, ClientInfo:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [SERVER]: finish startOfTurnUpdate
2020/12/20 16:09:28 [SERVER]: start runOrgs
2020/12/20 16:09:28 [SERVER]: start runIIGO
2020/12/20 16:09:28 [SERVER]: finish runIIGO
2020/12/20 16:09:28 [SERVER]: start runIIFO
2020/12/20 16:09:28 [SERVER]: finish runIIFO
2020/12/20 16:09:28 [SERVER]: start runIITO
2020/12/20 16:09:28 [SERVER]: finish runIITO
2020/12/20 16:09:28 [SERVER]: finish runOrgs
2020/12/20 16:09:28 [SERVER]: start endOfTurn
2020/12/20 16:09:28 [SERVER]: start runOrgsEndOfTurn
2020/12/20 16:09:28 [SERVER]: start runIIGOEndOfTurn
2020/12/20 16:09:28 [SERVER]: finish runIIGOEndOfTurn
2020/12/20 16:09:28 [SERVER]: start runIIFOEndOfTurn
2020/12/20 16:09:28 [SERVER]: finish runIIFOEndOfTurn
2020/12/20 16:09:28 [SERVER]: start runIITOEndOfTurn
2020/12/20 16:09:28 [SERVER]: finish runIITOEndOfTurn
2020/12/20 16:09:28 [SERVER]: finish runOrgsEndOfTurn
2020/12/20 16:09:28 [SERVER]: start probeDisaster
2020/12/20 16:09:28 [SERVER]: finish probeDisaster
2020/12/20 16:09:28 [SERVER]: start incrementTurnAndSeason
2020/12/20 16:09:28 [SERVER]: finish incrementTurnAndSeason
2020/12/20 16:09:28 [SERVER]: start deductCostOfLiving
2020/12/20 16:09:28 [SERVER]: finish deductCostOfLiving
2020/12/20 16:09:28 [SERVER]: start updateIslandLivingStatus
2020/12/20 16:09:28 [SERVER]: finish updateIslandLivingStatus
2020/12/20 16:09:28 [SERVER]: finish endOfTurn
2020/12/20 16:09:28 [SERVER]: finish runTurn
2020/12/20 16:09:28 [SERVER]: start runTurn
2020/12/20 16:09:28 [SERVER]: TURN: 2, Season: 1
2020/12/20 16:09:28 [SERVER]: start startOfTurnUpdate
2020/12/20 16:09:28 [Team1]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x2, ClientInfo:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team5]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x2, ClientInfo:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team4]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x2, ClientInfo:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team2]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x2, ClientInfo:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team6]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x2, ClientInfo:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:28 [Team3]: Received start of turn game state update: gamestate.ClientGameState{Season:0x1, Turn:0x2, ClientInfo:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}
2020/12/20 16:09:29 [SERVER]: finish startOfTurnUpdate
2020/12/20 16:09:29 [SERVER]: start runOrgs
2020/12/20 16:09:29 [SERVER]: start runIIGO
2020/12/20 16:09:29 [SERVER]: finish runIIGO
2020/12/20 16:09:29 [SERVER]: start runIIFO
2020/12/20 16:09:29 [SERVER]: finish runIIFO
2020/12/20 16:09:29 [SERVER]: start runIITO
2020/12/20 16:09:29 [SERVER]: finish runIITO
2020/12/20 16:09:29 [SERVER]: finish runOrgs
2020/12/20 16:09:29 [SERVER]: start endOfTurn
2020/12/20 16:09:29 [SERVER]: start runOrgsEndOfTurn
2020/12/20 16:09:29 [SERVER]: start runIIGOEndOfTurn
2020/12/20 16:09:29 [SERVER]: finish runIIGOEndOfTurn
2020/12/20 16:09:29 [SERVER]: start runIIFOEndOfTurn
2020/12/20 16:09:29 [SERVER]: finish runIIFOEndOfTurn
2020/12/20 16:09:29 [SERVER]: start runIITOEndOfTurn
2020/12/20 16:09:29 [SERVER]: finish runIITOEndOfTurn
2020/12/20 16:09:29 [SERVER]: finish runOrgsEndOfTurn
2020/12/20 16:09:29 [SERVER]: start probeDisaster
2020/12/20 16:09:29 [SERVER]: finish probeDisaster
2020/12/20 16:09:29 [SERVER]: start incrementTurnAndSeason
2020/12/20 16:09:29 [SERVER]: finish incrementTurnAndSeason
2020/12/20 16:09:29 [SERVER]: start deductCostOfLiving
2020/12/20 16:09:29 [SERVER]: finish deductCostOfLiving
2020/12/20 16:09:29 [SERVER]: start updateIslandLivingStatus
2020/12/20 16:09:29 [SERVER]: finish updateIslandLivingStatus
2020/12/20 16:09:29 [SERVER]: finish endOfTurn
2020/12/20 16:09:30 [SERVER]: finish runTurn
2020/12/20 16:09:30 [SERVER]: Max turns '2' reached or exceeded
===== GAME CONFIGURATION =====
config.Config{MaxSeasons:0x64, MaxTurns:0x2, InitialResources:100, CostOfLiving:10, MinimumResourceThreshold:5, MaxCriticalConsecutiveTurns:0x3}
===== START OF TURN 1 (END OF TURN 0) =====
gamestate.GameState{Season:0x1, Turn:0x1, ClientInfos:map[shared.ClientID]gamestate.ClientInfo{Team1:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 2 (END OF TURN 1) =====
gamestate.GameState{Season:0x1, Turn:0x2, ClientInfos:map[shared.ClientID]gamestate.ClientInfo{Team1:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 3 (END OF TURN 2) =====
gamestate.GameState{Season:0x1, Turn:0x3, ClientInfos:map[shared.ClientID]gamestate.ClientInfo{Team1:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
2020/12/20 16:09:30 Writing JSON output to 'C:/Users/lhlee/Documents/SOMAS/SOMAS2020/output/output.json'
2020/12/20 16:09:30 Finished writing JSON output to 'C:/Users/lhlee/Documents/SOMAS/SOMAS2020/output/output.json'
```

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

File: internal/common/gamestate/clientgamestate.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package gamestate
+
+// ClientGameState contains game state only for a specific client.
+type ClientGameState struct {
```
Comment (preetl) [inline internal/common/gamestate/clientgamestate.go:4]: Good shout making a separate ClientGameState struct. So we do still have the GameState with all the ClientInfos for server funcs?

Comment (lhl2617) [reply internal/common/gamestate/clientgamestate.go:4]: yes :)

Comment (preetl) [reply internal/common/gamestate/clientgamestate.go:4]: so when we update an islands individual resource pool, do we have to update it in both places?

Comment (lhl2617) [reply internal/common/gamestate/clientgamestate.go:4]: nope, you only update it in the server's own `gameState`. When you call `s.gameStateUpdate`, it copies it over to the corresponding `ClientGameState` :)

Comment (lhl2617) [reply internal/common/gamestate/clientgamestate.go:4]: More info:

You get the corresponding `ClientGameState` by doing for example for team 1:
```
s.gameState.GetClientGameStateCopy(shared.Team1)
```
This gets a copy based on `s.gameState`.

Comment (preetl) [reply internal/common/gamestate/clientgamestate.go:4]: oh sweet!

**Review by preetl — APPROVED** [review]

**Pull request merged**